### PR TITLE
Add health check support for real servers

### PR DESCRIPTION
**Pull request by jsravn**
Health checks are set on the virtual service using a URL like
"http://:8080/health". Once set, the reconciler will start performing
periodic health checks against all the real server IPs. Any server that
exceeds the down threshold will have its weight set to 0. When it
exceeds the up threshold, the weight is restored to its previous value.

Further caveats:
* This is done on a local level - so each merlin instance tests its
connectivity to the real servers.
* Once a health check is set, all servers are considered to be down
until they pass their health check. This facilitates adding new servers
before they are ready to receive traffic.
* The implication of the prior two points is that on restart merlin won't
forward traffic until health checks have passed. This is similar to
as if the reconciler had not run yet on start up.

### CODE REVIEW

File: e2e/meradm/meradm_test.go [hunk 1]
```diff
@@ -82,6 +82,19 @@ var _ = Describe("Meradm", func() {
 			_, err = meradmErrored("service", "add", "service1")
 			Expect(err).To(HaveOccurred(), "should have failed validation")
 		})
+
+		It("can set a healthcheck", func() {
+			meradm("service", "add", "service1", "tcp", "10.1.1.1:888", "-s=wrr", "-b=flag-1,flag-2")
+			meradm("service", "edit", "service1",
+				"--health-endpoint=http:556/health", "--health-period=5s", "--health-timeout=1s",
```
Comment (jsravn) [inline e2e/meradm/meradm_test.go]: This is wrong, the endpoint needs to be a valid url eg `http://:556/health`. Need validation in the server for this, probably.

File: reconciler/healthchecks/healthchecks.go [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+	c.Lock()
+	defer c.Unlock()
+
+	info, ok := c.infos[id]
+	if !ok || info.healthCheck == nil {
+		log.Info("return nil")
+		return nil
+	}
+
+	log.Infof("checking down servers in %v", info)
```
Comment (jsravn) [inline reconciler/healthchecks/healthchecks.go]: remove comment

File: reconciler/healthchecks/healthchecks.go [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+	return &checker{infos: make(map[string]*checkInfo)}
+}
+
+func (c *checker) GetDownServers(id string) []string {
+	c.Lock()
+	defer c.Unlock()
+
+	info, ok := c.infos[id]
+	if !ok || info.healthCheck == nil {
+		log.Info("return nil")
```
Comment (jsravn) [inline reconciler/healthchecks/healthchecks.go]: remove comment

File: reconciler/healthchecks/healthchecks.go [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+
+	if check == nil {
+		return
+	}
+
+	// kick off new health checks with the updated check
+	for server, status := range info.serverStatuses {
+		stopCh := make(chan struct{})
+		info.serverStopChs[server] = stopCh
+		go status.healthCheck(stopCh, server, check)
```
Comment (jsravn) [inline reconciler/healthchecks/healthchecks.go]: refactor duplicate code with that in `AddServer`

File: reconciler/healthchecks/healthchecks.go [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+func (c *checker) Stop() {
+	c.Lock()
+	defer c.Unlock()
+
+	for _, info := range c.infos {
+		c.updateHealthCheck(info, nil)
+	}
+}
+
+// healthCheck the servers and update their status
```
Comment (jsravn) [inline reconciler/healthchecks/healthchecks.go]: singular not plural

File: reconciler/healthchecks/healthchecks.go [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+	timeout, err := ptypes.Duration(healthCheck.Timeout)
+	if err != nil {
+		panic(err)
+	}
+	client := http.Client{
+		Transport: tr,
+		Timeout:   timeout,
+	}
+
+	serverURL, err := url.Parse(fmt.Sprintf("http://%s:%s%s", server, checkURL.Port(), checkURL.Path))
```
Comment (jsravn) [inline reconciler/healthchecks/healthchecks.go]: Is there a nicer way to do this using the existing checkURL?

File: reconciler/healthchecks/healthchecks.go [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+type healthState bool
+
+const (
+	serverUp   healthState = true
+	serverDown healthState = false
+)
+
+type checkStatus struct {
+	state healthState
+	count uint32
```
Comment (jsravn) [inline reconciler/healthchecks/healthchecks.go]: comment that count is a variable of the state, either serverUp or serverDown state.

File: reconciler/healthchecks/healthchecks_test.go [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+	})
+
+	Context("server is down", func() {
+		BeforeEach(func() {
+			setServerStatus(http.StatusInternalServerError)
+		})
+
+		It("should list server in down IPs", func() {
+			checker.SetHealthCheck(id, check)
+			checker.AddServer(id, "127.0.0.1")
```
Comment (jsravn) [inline reconciler/healthchecks/healthchecks_test.go]: Test multiple servers here.

File: reconciler/healthchecks/healthchecks.go (hunk 7 above)
Comment (alangibson01) [reply reconciler/healthchecks/healthchecks.go]: I think changing the variable name to something more descriptive would be better than adding a comment. Exactly what name would be better escapes me at the moment, however.

File: reconciler/reconciler.go [hunk 9]
```diff
@@ -111,6 +117,10 @@ func (r *reconciler) reconcile() {
 			}
 		}
 
+		if match == nil || !proto.Equal(desiredService.HealthCheck, match.HealthCheck) {
+			r.checker.SetHealthCheck(desiredService.Id, desiredService.HealthCheck)
```
Comment (alangibson01) [inline reconciler/reconciler.go]: Is it worth logging that a healthcheck is being set, similar to the above `Adding virtual service: %v` message?

File: reconciler/healthchecks/healthchecks.go (hunk 6 above)
Comment (jsravn) [reply reconciler/healthchecks/healthchecks.go]: It's a bit ugly mutating the previous url struct fields, so I've left this as is.

**Comment by jsravn** [comment]
@alangibson01 can you elaborate on how to expand the transition coverage? I thought I was already covering it fairly well.

**Comment by alangibson01** [comment]
@jsravn Some examples, with upThreshold and downThreshold set to values > 1:

* Service starts down, passes one healthcheck, then goes back to failing => should stay down.
* Service starts up, fails one healthcheck, then goes back to passing => should stay up.

I don't think you've got those cases covered, unless I've missed them.

File: reconciler/reconciler.go (hunk 9 above)
Comment (jsravn) [reply reconciler/reconciler.go]: I've added it to the health checker.

**Comment by jsravn** [comment]
@alangibson01 I added some additional coverage of the transitions.

I'm still struggling with why the build seems to be flaky on travis. It randomly can't seem to find the expanded tarball. It's bizarre.

File: reconciler/reconciler_test.go [hunk 10]
```diff
@@ -36,7 +39,7 @@ var _ = Describe("Reconciler", func() {
 			Flags:     []string{"flag-1", "flag-2"},
 		},
 		HealthCheck: &types.VirtualService_HealthCheck{
-			Endpoint:      "http:102/health",
+			Endpoint:      &wrappers.StringValue{Value: "http:102/health"},
```
Comment (alangibson01) [inline reconciler/reconciler_test.go]: Is this supposed to be invalid, or just an oversight?

**Comment by jsravn** [comment]
@alangibson01 It's really hard to test the transition state since it's opaque, and is sensitive to timings (when the checker runs exactly after the test starts). I think it's sufficient to test the checker runs at least n times (via waitForDown/waitForUp).

**Comment by alangibson01** [comment]
Re: transitions, I'm happy now that the up-down-up and down-up-down tests are in there.

File: cmd/meradm/service.go [hunk 11]
```diff
@@ -86,42 +86,38 @@ func init() {
...
-		if healthTimeout != 0 {
-			svc.HealthCheck.Timeout = ptypes.DurationProto(healthTimeout)
-		}
-		if healthUpThreshold != 0 {
-			svc.HealthCheck.UpThreshold = uint32(healthUpThreshold)
-		}
-		if healthDownThreshold != 0 {
-			svc.HealthCheck.DownThreshold = uint32(healthDownThreshold)
-		}
+	if cmd.Flag("health-endpoint").Changed {
```
[the remark this answers is not in the document]
Comment (jsravn) [reply cmd/meradm/service.go]: It's the only way to determine if a flag was set by a user. Not sure what you mean about using the `Flag`?

File: reconciler/reconciler_test.go (hunk 10 above)
Comment (jsravn) [reply reconciler/reconciler_test.go]: Health checker is mocked out so it doesn't matter what it's set to. But I can fix it.

File: cmd/meradm/service.go [hunk 12]
```diff
@@ -74,7 +74,7 @@ func init() {
 		f.StringVarP(&scheduler, "scheduler", "s", "", "scheduler for new connections")
 		f.StringSliceVarP(&schedFlags, "sched-flags", "b", nil, "scheduler flags")
 		f.StringVar(&healthEndpoint, "health-endpoint", "", "Endpoint for health checks. "+
-			"If unset, no health check occurs and the server is assumed to always be up.")
+			"If set to empty string, no health check occurs and the server is assumed to always be up.")
```
[the remark this answers is not in the document]
Comment (jsravn) [reply cmd/meradm/service.go]: Okay I'll update this in my next PR which already modifies these descriptions.

**Pull request merged**